### PR TITLE
Add an APM metric to aggregations usage

### PR DESCRIPTION
**Pull request by not-napoleon**
This wires up the "new" APM metrics integration to the existing Aggregations usage tracking system.  It introduces one new metric, a `LongCounter` named `es.search.query.aggregations.total`, which has dimensions for the specific aggregation being run, and the values source type we resolved it to.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/SearchModule.java [hunk 1]
```diff
@@ -352,7 +371,12 @@ public Map<String, Highlighter> getHighlighters() {
     }
 
     private ValuesSourceRegistry registerAggregations(List<SearchPlugin> plugins) {
-        ValuesSourceRegistry.Builder builder = new ValuesSourceRegistry.Builder();
+        ValuesSourceRegistry.Builder builder = new ValuesSourceRegistry.Builder(
+            telemetryProvider == null
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/search/SearchModule.java:375]: telemetry provider shouldn't be null. Have you seen cases where it is?

**Review by rjernst — COMMENTED** [review]
Looks fine to me, one nit. @ldematte might have other thoughts

File: server/src/main/java/org/elasticsearch/search/SearchModule.java (hunk 1 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java:375]: There's like 200 places in the tests where we create a `SearchModule`, as well as a couple places in production code where we do this: 

```
NamedWriteableRegistry registry = new NamedWriteableRegistry(new SearchModule(Settings.EMPTY, List.of()).getNamedWriteables());
```

(which seems like a bit of an anti-pattern, but also massively out of scope for this PR. 

At any rate, I defaulted it to null for those cases via a defaulting constructor (see line 315).  If there's something better than Null to default it to, I'm happy to do that instead, but I didn't see an obvious choice.

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java:375]: There is `TelemetryProvider.NOOP`

Comment (ldematte) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java:375]: ++
If you can use TelemetryProvider.NOOP, you could enforce it's always non-null and remove the null checks

**Review by ldematte — COMMENTED** [review]
Looks fine to me (and thanks for choosing a sensible name for the metric!)
As for testing, in case you want to assert/verify a metric is emitted, there is a `RecordingMeterRegistry` class you can use for that. For example, you can look how it is used by `HierarchyCircuitBreakerTelemetryTests`

File: server/src/main/java/org/elasticsearch/search/SearchModule.java (hunk 1 above)
Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java:375]: oh, spiff.  Not sure how I missed that.  I'll use that instead.

File: server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationUsageService.java [hunk 2]
```diff
@@ -61,6 +74,10 @@ public void incAggregationUsage(String aggregationName, String valuesSourceType)
             assert adder != null : "Unknown subtype [" + aggregationName + "][" + valuesSourceType + "]";
         }
         assert valuesSourceMap != null : "Unknown aggregation [" + aggregationName + "][" + valuesSourceType + "]";
+        // Also not registered during tests
+        if (aggregationsUsageCounter != null) {
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/AggregationUsageService.java:78]: noop meter registry will return noop counters. no need for null check

File: server/src/main/java/org/elasticsearch/search/SearchModule.java [hunk 3]
```diff
@@ -352,7 +371,10 @@ public Map<String, Highlighter> getHighlighters() {
     }
 
     private ValuesSourceRegistry registerAggregations(List<SearchPlugin> plugins) {
-        ValuesSourceRegistry.Builder builder = new ValuesSourceRegistry.Builder();
+        ValuesSourceRegistry.Builder builder = new ValuesSourceRegistry.Builder(
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/search/SearchModule.java:374]: how about we move the registration closer to the usage - so inside the AggregationUsageService.Builder ?

**Review by pgomulka — COMMENTED** [review]
it looks good, left 2 comments

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Review by nik9000 — APPROVED** [review]

**Review by iverase — APPROVED** [review]
LGTM

**Pull request merged**